src/module.c

```diff
@@ -6217,18 +6313,40 @@ RedisModuleCallReply *RM_Call(RedisModuleCtx *ctx, const char *cmdname, const ch
 
     user *user = NULL;
     if (flags & REDISMODULE_ARGV_RUN_AS_USER) {
-        user = ctx->user ? ctx->user->user : ctx->client->user;
+        if (ctx->client) {
+            user = ctx->client->user;
+        }
+        if (ctx->persistent_client) {
+            user = ctx->persistent_client->client->user;
+        } else if (ctx->user) {
+            user = ctx->user->user;
+        }
+
         if (!user) {
             errno = ENOTSUP;
             if (error_as_call_replies) {
                 sds msg = sdsnew("cannot run as user, no user directly attached to context or context's client");
                 reply = callReplyCreateError(msg, ctx);
             }
+            moduleFreeArgv(argv, argc);
             return reply;
         }
     }
 
-    c = moduleAllocTempClient(user);
+    if (ctx->persistent_client) {
+        c = ctx->persistent_client->client;
+    } else {
+        c = moduleAllocTempClient(user);
+    }
+
+    if (c->flags & CLIENT_BLOCKED) {
+        errno = -EADDRINUSE;
+        if (error_as_call_replies) {
+            sds msg = sdsnew("cannot reuse redis module client while blocked on another command");
+            reply = callReplyCreateError(msg, ctx);
+        }
+        goto cleanup;
+    }
```

Comment:
so that's an unrelated improvement, right? could have been part of the async RM_Call PR?

Reply:
it was to preventthis from being used in conjuction with async rm_call's as that would releaes the client itself, code has been improved to work with it now